### PR TITLE
Adds LCEL and Langserve support 

### PR DESCRIPTION
**Pull request by vmesel**
In this PR we are adding the possibility of loading a pure LCEL Runnable based on a code that the user can add to its docker instance, making it compatible with our current API responses and, also, adding the langserve support for LCEL.

In this LCEL implementation, we are still missing the semantic search with the vectors, that will be done later.

related #170 #124

### CODE REVIEW

**Comment by avelino** [comment]
i don't have the capacity to review this implementation, i would have to study LCEL in depth to be able to suggest any improvements to the talkd/dialog implementation - currently i don't have the time for this in-depth study as i would like to 

sorry I can't contribute to this PR

**Comment by vmesel** [comment]
@avelino I partially disagree that you can't add value here, LCEL is not the total focus here, we changed some functions and signatures to allow users to import RAGs made with LCEL.

The only LCEL file we have is a new file for a sample RAG.

File: src/dialog/llm/__init__.py [hunk 1]
```diff
@@ -20,5 +26,40 @@ def get_llm_class() -> Type[AbstractLLM]:
...
+            {"input": message.message},
+            {"configurable": {
+                "session_id": chat_id,
+                "model": Settings().PROJECT_CONFIG.get("model_name", "gpt-3.5-turbo"),
+                **Settings().PROJECT_CONFIG,
+            }}
+        )
+        ai_message = {"text": ai_message.content}
+
+    return ai_message
```
Comment (lgabs) [inline src/dialog/llm/__init__.py:59]: why does this function lives here in the `__init__.py`? If each class implementation have its own methods for processing, why not leave the chain with LCEL inside the `DialogLcelLLM` and make its parts more LCEL-like (prompt, retriever, memory etc)?

File: src/dialog/llm/__init__.py [hunk 2]
```diff
@@ -20,5 +26,40 @@ def get_llm_class() -> Type[AbstractLLM]:
         llm_class = getattr(llm_module, class_name)
     except Exception as e:
         logging.info(f"Failed to load LLM class {Settings().LLM_CLASS}. Using default LLM. Exception: {e}")
+        raise
+
+    if isinstance(llm_class, AbstractLLM):
+        return llm_class, "AbstractLLM"
+
+    elif 'langchain_core.runnables' in str(type(llm_class)):
+        return llm_class, "LCELRunnable"
```
Comment (lgabs) [inline src/dialog/llm/__init__.py:35]: why not check instance types directly instead of returning one more variable just for string comparison? For example, runnables can be checked like (using the `RunnableWithMessageHistory` defined in your `lcel.py` implementation)
```
In [76]: issubclass(RunnableWithMessageHistory, Runnable)
Out[76]: True
```

File: src/dialog/llm/__init__.py [hunk 3]
```diff
@@ -20,5 +26,40 @@ def get_llm_class() -> Type[AbstractLLM]:
...
+    return DialogLLM, "AbstractLLM"
+
+
+def process_user_message(message, chat_id=None):
+    LLM, llm_type = get_llm_class()
+    if llm_type == "AbstractLLM":
+        llm_instance = LLM(config=Settings().PROJECT_CONFIG, session_id=chat_id)
+        ai_message = llm_instance.process(message.message)
+
+    elif llm_type == "LCELRunnable":
```
Comment (lgabs) [inline src/dialog/llm/__init__.py:48]: could it be easier to maintain the previous DialogLcelLLM and make its interface gradually equal to a langchain runnable, avoiding this kind of handler? This seems to create more complexity to find a common ground sometime.

As an example of implementation, for LCEL wrapper, configurables and other configuration parameters would be initialized in the init, process method could call a chain's invoke method, and the chain could be fully constructed with runnables For example
```
class DialogLcelLLM(AbstractLLM):
    def __init__(self, ...):
        super().__init__(...)
        # handle init parameters in each class case
     def generate_prompt(...):
     # creates a ChatPromptTemplate 
     def llm(...)
     # the llm component
     def chain(...)
     # this is new, could represent better the LCEL chain using the correct name
     def process(...)
     # this takes the input question and processes, in LCEL case it woudl call self.chain.invoke(inputs)
     def postprocess(...)
     # custom postprocessing
```

File: src/dialog/llm/agents/lcel.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    | prompt
+    | chat_model
+)
+
+runnable = RunnableWithMessageHistory(
+    chain,
+    get_memory_instance,
+    input_messages_key='input',
+    history_messages_key="chat_history"
+)
```
Comment (lgabs) [inline src/dialog/llm/agents/lcel.py:78]: nice example of a RAG Chain :) Just a note: most examples use an additional step to summarize history+question in a standalone question to make a better instruction for the llm, but since it's the current behavior we could leave it like this and compare better later.

File: src/dialog/llm/agents/lcel.py
Comment (lgabs) [inline src/dialog/llm/agents/lcel.py:1]: I think this naming can confuse a lot developers working with langchain. The chain implementation with LCEL is actually a _chain_ from langchain, which is built with LCEL. So, the main output of `lcel.py` is a _chain_, I think we could call it for what it is. 

In this direction, `agents` folder can confuse as well, because it's holding _chains_ and a wrapper for chains (the `DialogLLM` class in `default.py`), and not [langchain's agents](https://python.langchain.com/v0.1/docs/modules/agents/how_to/custom_agent/#create-the-agent). Yet, the `agent` folder is inside `llm` folder, when usually it's the inverse: the agent uses one or more llms. 

The `llm` folder is indeed a folder for _chain wrappers_ (our subclasses of AbstractLLM), and **In my opinion we have to chose** between keeping these wrappers (in this case we can wrapper chains using LCEL as [shown here](https://github.com/talkdai/dialog/pull/185/files#r1597521342) without adding much complexity ) or, if we want to abandon them, keep a whole simpler separate folder for chains (e.g.,a `chain` folder with one file per chain). It seems to me that, without this, complexity will raise a lot to route correctly between these patterns

**Comment by lgabs** [comment]
I'd like to take the opportunity here to point something important: the taxonomy we use when speaking about LLM Applications is important for our communication. There are many confusing terms and many recent concepts, so I just want to help us to use the same names for the same things:

- _LLM (Large Language Models)_: this is the model per-se, not the chains, not an agent. We sometimes use _llm_ to denote a _chain_ that calls an _llm_ (even langchain's example call instances of chains by `llm` in this case), but it's important to us to keep this in mind.
- [LangChain Expression Language](https://python.langchain.com/v0.1/docs/expression_language/), or LCEL, is a declarative way to easily compose chains together. It's the a way to build chains, so we could avoid calling chains by this name.
- _Chains_ are a more abstract concept from langchain (it's its base concept), of [_"Chains refer to sequences of calls - whether to an LLM, a tool, or a data preprocessing step"_](https://python.langchain.com/v0.1/docs/modules/chains/). They are built preferably with LCEL joining one or more runnables.
- [_Runnables_](https://python.langchain.com/v0.1/docs/expression_language/interface/): the new protocol used by most of langchain components, making it easier to compose chains like `runnable1 | runnable2`. It's kind of okay to call a chain by _runnable_ since it is its subclass, but the name makes it easier to understand that chains are compositions, and runnables are components. For example, langchain's prompt templates, memory components (like RunnableWithMessageHistory) and retrievers (like PGVector) are also runnables.  
- _RAG_: RAG is a technique for augmenting LLM knowledge with additional data. It's just the technique, so it'd not be good to call chains or runnables by RAG. A chain _uses_ RAG technique to enhance it's performance and knowledge of the world.
- _LLM Applications_: BTW this is a better name for what dialog project produces: an application that uses LLM to power is capability. It's not _only_ a _LLM_, but a whole infrastructure around it.

File: src/dialog/llm/agents/lcel.py
Comment (vmesel) [reply src/dialog/llm/agents/lcel.py:1]: I think we are just using Langchain names here. According to other sources:

An agent is:

> LLM Agents are advanced AI systems that use LLMs to understand and generate human language, in context and in a sophisticated manner. LLM Agents go beyond simple text generation. They can maintain the thread of a conversation, recall previous statements, and adjust their responses accordingly with different tones and styles.
Source: https://www.iguazio.com/glossary/llm-agents/

A chain is:

>An LLM Chain, short for Large Language Model Chain, is a powerful concept within the LangChain framework that combines different primitives and large language models (LLMs) to create a sequence of operations for natural language processing (NLP) tasks such as completion, text generation, text classification, etc. It allows to orchestrate a series of steps to process user input, generate prompts (using PromptTemplate), and obtain meaningful outputs from an LLM. 
Source: https://vitalflux.com/llm-chain-openai-chatgpt-python-example/

In Langchain terms, we are just returning chains, but that's not the naming across other languages. I would agree to name the folder `chains` if we didn't have a wrapper class and we were returning just chained objects across all implementations.

We can start a discussion on naming, but I don't want to make this longer in this PR, letting us focus on the LCEL implementation only.

----

TL;DR: Langchain naming and our naming is different. We need some common ground over here, so we need to open up a discussion on the naming.

I'm not fixing the naming conventions in this PR to make it small and concise only to serve the purpose of allowing users to use LCEL on Dialog.

File: src/dialog/llm/__init__.py (hunk 3 above)
Comment (vmesel) [reply src/dialog/llm/__init__.py:48]: Well, it would have been easier in the past, before this whole implementation hahahaha. If we do agree to this common ground, I can rollback without any problems.

IMHO, I think we need to have support for both structures, allowing users to easily migrate to our solution, but that's what I think.

File: src/dialog/llm/__init__.py (hunk 1 above)
Comment (vmesel) [reply src/dialog/llm/__init__.py:59]: This code was made in the __init__ so it's accessible through the module and on the API across different views. The theory is that we need to make thin views and fat models, but since we don't have too many models or business logic that runs on models, we should use this function as a controller, which process something and shrink the size of the view.

File: src/dialog/llm/__init__.py (hunk 2 above)
Comment (vmesel) [reply src/dialog/llm/__init__.py:35]: I need to check on the computer why I've done this way.

File: src/dialog/llm/agents/lcel.py
Comment (lgabs) [reply src/dialog/llm/agents/lcel.py:1]: Ok, naming conventions can be discussed later, I've left some comments about that because I find it very important for communication before we establish many patterns which can confuse with langchain's ones.

**Comment by vmesel** [comment]
> I'd like to take the opportunity here to point something important: the taxonomy we use when speaking about LLM Applications is important for our communication. There are many confusing terms and many recent concepts, so I just want to help us to use the same names for the same things:
> 
> * _LLM (Large Language Models)_: this is the model per-se, not the chains, not an agent. We sometimes use _llm_ to denote a _chain_ that calls an _llm_ (even langchain's example call instances of chains by `llm` in this case), but it's important to us to keep this in mind.
> * [LangChain Expression Language](https://python.langchain.com/v0.1/docs/expression_language/), or LCEL, is a declarative way to easily compose chains together. It's the a way to build chains, so we could avoid calling chains by this name.
> * _Chains_ are a more abstract concept from langchain (it's its base concept), of [_"Chains refer to sequences of calls - whether to an LLM, a tool, or a data preprocessing step"_](https://python.langchain.com/v0.1/docs/modules/chains/). They are built preferably with LCEL joining one or more runnables.
> * [_Runnables_](https://python.langchain.com/v0.1/docs/expression_language/interface/): the new protocol used by most of langchain components, making it easier to compose chains like `runnable1 | runnable2`. It's kind of okay to call a chain by _runnable_ since it is its subclass, but the name makes it easier to understand that chains are compositions, and runnables are components. For example, langchain's prompt templates, memory components (like RunnableWithMessageHistory) and retrievers (like PGVector) are also runnables.
> * _RAG_: RAG is a technique for augmenting LLM knowledge with additional data. It's just the technique, so it'd not be good to call chains or runnables by RAG. A chain _uses_ RAG technique to enhance it's performance and knowledge of the world.
> * _LLM Applications_: BTW this is a better name for what dialog project produces: an application that uses LLM to power its capability. It's not _only_ a _LLM_, but a whole infrastructure around it.

I will move this to an issue for later discussion.

File: src/dialog/llm/__init__.py (hunk 3 above)
Comment (vmesel) [reply src/dialog/llm/__init__.py:48]: @lgabs what do you think: let's let the user use this structure as it is and then we implement the class in a future PR (probably this week, getting the DialogLcelLLM class back), or do you think this is a no go in this PR?

**Comment by avelino** [comment]
omg, a lot of discussions indirectly connected to this PR, it makes sense to open up some discussions to talk about specific points that came up in this PR.

**Review by avelino — APPROVED** [review]
I've looked at the implementation and it's a great start for RAG architecture following the use of langserver and its internals 

LGTM

we'll probably have changes in the future due to the points @lgabs made in his comments above
